### PR TITLE
[Bug-bash] Refine validations for tracing imperative APIs

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11562?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11562/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11562
```

</p>
</details>

### What changes are proposed in this pull request?

This PR fixes a few papercuts about state validation and error messages for MLflow Client APIs for trace creation.

**1. Disallow calling `MlflowClient().start_trace()` when there is an active span created by fluent APIs.**
The `start_trace()` method tries to create a root span, however, if there is active span created by fluent API, the new span will be child span of it. We block doing that and instead recommend customers to avoid mixing two APIs.
```
with mlflow.start_span(xxx):
    MlflowClient().start_trace(yyy) <- This should raise error
```

**2. Raise exception when `MlflowClient().end_span()` is called for already ended span_id.**

This only emit warning currently, but should be an exception as it is invalid use of API not internal error.

**3. Fix error message when `MlflowClient().end_trace()` is called for already ended request_id.**
When `end_trace` is accidentally called twice, we raises an error.
```
root_span = MlflowClient().start_trace()
MlflowClient().end_trace(root_span.request_id)
MlflowClient().end_trace(root_span.request_id)
```
Current error message is `Trace with ID xxx not found`, which is not accurate for this case. This PR changes it to `Trace with ID xxx already finished` by adding validation to check if the trace is already recorded or not.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/trace_manager.py [hunk 1]
```diff
@@ -154,9 +153,6 @@ def get_span_from_id(self, request_id: str, span_id: str) -> Optional[MLflowSpan
 
         return trace.span_dict.get(span_id) if trace else None
 
-    # NB: Caching as this requires a linear search over all spans in the trace and
```
Comment (B-Step62) [inline mlflow/tracing/trace_manager.py:157]: Note: Removing caching as it caused issue when `end_trace()` API is called twice. When a trace is popped out, this method should return None but the cache doesn't reflect it. The number of span in a single trace should be at most a few hundreds in normal use case and this method is only called within `end_trace()`, so this shouldn't be a big concern.

File: mlflow/tracking/client.py [hunk 2]
```diff
@@ -485,6 +487,24 @@ def start_trace(
...
         """
+        # Validate no active trace is set in the global context. If there is an active trace,
+        # the span created by this method will be a child span under the active trace rather than
+        # a root span of a new trace, which is not desired behavior.
+        if span := mlflow.get_current_active_span():
+            raise MlflowException(
+                f"Another trace is already set in the global context with ID {span.request_id}. "
+                "It appears that you have already started a trace using fluent APIs like "
+                "`@mlflow.trace()` or `with mlflow.start_span()`. However, it is not allowed "
+                "to call MlflowClient.start_trace() under an active trace created by fluent APIs, "
```
Comment (BenWilson2) [inline mlflow/tracking/client.py:498]:
```suggestion
                "to call MlflowClient.start_trace() under an active trace created by fluent APIs "
```

File: mlflow/tracing/clients/local.py [hunk 3]
```diff
@@ -72,7 +75,23 @@ def get_traces(self, n: int = 10) -> List[Trace]:
         Returns:
             A list of Trace objects.
         """
-        return list(self.queue) if n is None else list(self.queue)[-n:]
+        with self._lock:
+            trace_list = list(self.queue)
+        return trace_list if n is None else trace_list[-n:]
```
Comment (harupy) [inline mlflow/tracing/clients/local.py:80]: Can we update the type hint for `n`?

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -496,6 +498,24 @@ def start_trace(
...
+        # a root span of a new trace, which is not desired behavior.
+        if span := mlflow.get_current_active_span():
+            raise MlflowException(
+                f"Another trace is already set in the global context with ID {span.request_id}. "
+                "It appears that you have already started a trace using fluent APIs like "
+                "`@mlflow.trace()` or `with mlflow.start_span()`. However, it is not allowed "
+                "to call MlflowClient.start_trace() under an active trace created by fluent APIs "
+                "because it may lead to unexpected behavior. To resolve this issue, consider the "
+                "following options:\n"
+                " - If you want to create a child span under the active trace, use "
```
Comment (harupy) [inline mlflow/tracking/client.py:512]:
```suggestion
                " - To create a child span under the active trace, use "
```

I'd remove `If you want`.

File: mlflow/tracing/clients/base.py [hunk 5]
```diff
@@ -5,10 +5,11 @@
 
 class TraceClient(ABC):
     @abstractmethod
-    def log_trace(self, trace: Trace):
+    def get_trace(self, request_id: Trace):
```
Comment (serena-ruan) [inline mlflow/tracing/clients/base.py:8]:
```suggestion
    def get_trace(self, request_id: str):
```

**Pull request merged**